helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -200,9 +200,10 @@ private void generateMessage(final Resource resource, final BaseControllerDataPr
         }
 
         for (Message staleMessage : staleMessages) {
-          if (System.currentTimeMillis() - currentStateOutput
+          if ((System.currentTimeMillis() - currentStateOutput
               .getEndTime(resourceName, partition, instanceName)
-              > DEFAULT_OBSELETE_MSG_PURGE_DELAY) {
+              > DEFAULT_OBSELETE_MSG_PURGE_DELAY)
+              && staleMessage.getResourceName().equals(resourceName)) {
```

Comment:
Since this logic is under per-resource, per-partition and per-instance condition, we should also check whether the staleMessage's corresponding partition matches the current `partition`, on top of checking resource. Otherwise, if the instance has multiple partitions of this resource, this log will still be duplicate.